### PR TITLE
add resource module

### PR DESCRIPTION
**Pull request by garlick**
This is a bare bones starting point for a resource module as discussed in #2908.

It just tries to look up `resource.hwloc.by_rank`.  If that succeeds, do nothing.  If that fails, wait for all brokers to come online according to `hello.idset`, then run `flux hwloc reload` as a subprocess.  Since the schedulers both look up `resource.hwloc` keys using FLUX_KVS_WAITCREATE before they say HELLO to the job manager, they just work as before.

This allows `flux hwloc reload` to be taken out of rc1 so that rc1 on rank 0 can start up before the other brokers are online in a future PR.

I named the module `resource` but temporarily register its name internally with `MOD_NAME ("xresource")`, until flux-framework/flux-sched#485 is addressed.

I'm not sure if this is worthy of being merged on its own, but I can at least rebase the rc startup changes on this to make progress.

### CODE REVIEW

**Comment by grondo** [comment]
We will need a resource module at some point anyway, so perhaps this point is moot, but would we want a script that runs on rank 0 once all other ranks are "up"? i.e. the reverse of the instance "epilog" we were discussing before? If so, another way to run the initial `flux hwloc reload` would be via this script. 

(I realize this `resource` module is just a stand-in until we can get a real resource module in, so perhaps my comment was meaningless. I was just thinking that current functionality might be useful more generally if, instead of running a specific command, a configurable script was run...)

**Comment by garlick** [comment]
Hmm, could we accomplish the same thing with an rc2 script that runs the initial program as a wrapper?  E.g.
```sh
if test $rank  -eq 0; then
    flux hello --wait-for all
    flux hwloc reload

    # run user's initial program, if any, (somehow) and capture exit code for broker exit

    flux queue stop --quiet
    flux job cancelall --user=all --quiet -f --states RUN
    flux queue idle --quiet
fi
```

**Comment by grondo** [comment]
> Hmm, could we accomplish the same thing with an rc2 script that runs the initial program as a wrapper? E.g.

That seems like it could work. There may be tradeoffs either way. For example an rc2 script that embeds the user's initial program may be more tricky to customize, whereas separating rc2 from an rc1-post or epilog might be easier to mix and match. 

However, it isn't clear now (at least to me) what the right approach might be, so I'd be fine with whatever is easiest.

**Comment by garlick** [comment]
Not clear to me either.

**Comment by garlick** [comment]
Restarted a builder that failed here:
```
expecting success: 
	flux start --caliper-profile=thread-trace /bin/true

Segmentation fault (core dumped)
not ok 1 - --caliper-profile works

expecting success: 
	test -f "$CALIPER_OUTPUT"

not ok 2 - caliper output file exists

expecting success: 
	cali-query "$CALIPER_OUTPUT"

cali-query: Error: Could not read file *.cali
__rec=globals
ok 3 - caliper output file is readable by cali-query
caliper output file is readable by cali-query

# failed 2 among 3 test(s)
```

**Comment by garlick** [comment]
Just pushed an update.  There's quite a bit more here now, with a basic sharness test.
Summary of module capability is in the commit message of 1335002498f.

**Comment by garlick** [comment]
The model proposed here thus far is for `resource` module to configure/discover/track resources and emit events to `resource.eventlog` for scheduler to consume (possibly via libschedutil helpers). 

In an offline discussion with @grondo, we thought it would be more efficient for `resource` to provide the current state of resources to the scheduler, rather than requiring it to replay an ever-growing eventlog on startup.  The eventlog is still useful so this would be in addition to what's there now.

Here's a straw man proposal for a protocol:

**resource.reserve**

* Streaming RPC
* Request message currently empty (interpreted as "reserve all"), but eventually could include an idset to reserve a configured subset of resources
* Scheduler doesn't start HELLO with job-manager until first response is received
* first response includes resources in some to-be-defined format (initially let's use the `hwloc.by_rank` object, with the understanding that flux-sched can use its hwloc reader to ingest hwloc.xml.* as before once this response is received).
* After first response, satisfiability of job requests can start because the "base set" of resources are known.
* Scheduler considers all exec targets "unavailable" at first
* As exec targets come online, more responses are generated

Responses would be an object with the following optional top level keys:
* `resources`: initial R or by_rank object defining resources explicitly
* `exclude`: idset of exec target ranks to exclude from scheduling
* `available`: idset of exec target ranks that are online and ready for work
* `drain`: idset of exec target ranks that should no longer be considered for jobs

The first response would include all keys and subsequent responses would include only those keys whose value has changed.

**Comment by grondo** [comment]
> `resource.reserve`

BTW, I liked your term "acquire" that you used on today's call. Would that be more appropriate than "reserve" for this first scheduler RPC?

**Comment by dongahn** [comment]
One thing that came out at today's call:

> Scheduler doesn't start HELLO with job-manager until first response is received

We may not have to tie the resource synchronization with job-manager synchronization? In the case of `sched-simple` with a single-component architecture, both synchronizations need be done by one module. But In the case of flux-sched with a two-component architecture,  `fluxion-resource` will be responsible for resource synchronization while `qmanager` will be responsible for `job-manager` synchronization. 

So `qmanager`-`job-manager` sync doesn't have to depend on the resource synchronization... (`qmanager` and `fluxion-resource` sync have to happen -- but that'd be internal to flux-sched.)

**Comment by garlick** [comment]
> We may not have to tie the resource synchronization with job-manager synchronization? 

Yeah, I think I would rephrase my statement to something like "Scheduler does not evaluate job request satisfiability until first response is received".   qmanager could run HELLO and start accepting alloc requests, as long as it doesn't reject them before acquiring resources.

**Comment by garlick** [comment]
Asan builder failed with a kill -9 timeout on `t2611-debug-emulate.t`.  Restarted.

**Comment by garlick** [comment]
Hmm, some tests added in this PR are failing in the `Ubuntu: py3.7, clang-6.0 chain-lint --with-flux-security` builder.  I will need to run that down.
```
not ok 14 - undrain fails if rank not drained
FAIL: t2310-resource-module.t 14 - undrain fails if rank not drained
#	
#		test_must_fail undrain_idset 1 2>undrain_not.err &&
#		grep "rank 1 not drained" undrain_not.err
#	
not ok 15 - drain fails if idset is malformed
FAIL: t2310-resource-module.t 15 - drain fails if idset is malformed
#	
#		test_must_fail drain_idset_noreason xxzz 2>drain_badset.err &&
#		grep "failed to decode idset" drain_badset.err
#	
not ok 16 - drain fails if idset is empty
FAIL: t2310-resource-module.t 16 - drain fails if idset is empty
#	
#		test_must_fail drain_idset_noreason "" 2>drain_empty.err &&
#		grep "idset is empty" drain_empty.err
#	
not ok 17 - drain fails if idset is out of range
FAIL: t2310-resource-module.t 17 - drain fails if idset is out of range
#	
#		test_must_fail drain_idset_noreason "0-$SIZE" 2>drain_range.err &&
#		grep "idset is out of range" drain_range.err
#	
```

**Comment by garlick** [comment]
Per offline discussion with @grondo, just force pushed a squash of the fixups, and a protocol change - replacing "available" idset with complementary "up" and "down" idsets.

**Comment by garlick** [comment]
Restarted a builder that failed here
```
ERROR: lua/t0002-rpc.t - exited with status 137 (terminated by signal 9?)
```

**Comment by garlick** [comment]
pushed a fixup for a missing HAVE_JQ in the sharness test

**Comment by garlick** [comment]
I think this may be ready for a review.

**acquire RPC**
We iterated on the `resource.acquire` RPC:

First response:
```
{resources:resource_object up:idset}
```
Subsequent responses:
```
{up?:idset down?:idset}
```
Where `resource_object` is currently the `hwloc.by_rank` object, but would be replaced a new design as discussed above.  The `up` idset, and its later modification by `up`/`down` which are deltas, not replacement sets, encapsulates all changes to availability.  Excluded execution targets are not included in the initial resource object with the idea that this set is to be used for satisfiability.  There is some more detailed info in the block comment at the top of [`resource/acquire.c` ](https://github.com/garlick/flux-core/blob/resource_module/src/modules/resource/acquire.c)

**resource.eventlog**
The eventlog is now used just for recovery/history purposes.  When the resource module loads, it post a `resource-init` event that summarizes its starting state, then an event is posted for each change thereafter.  For restart purposes, all the events leading up to the last `resource-init` event can be ignored.  Here's an example eventlog where the module was first loaded, then reloaded:
```
1589944084.927766 resource-init {"restart":false,"hwloc-discover":false,"online":"0-3","drain":"","exclude":""}
1589944084.932976 hwloc-discover-start {"pid":23975}
1589944085.277061 hwloc-discover-finish {"loaded":true}
1589944103.970201 drain {"idset":"0","reason":""}
1589944114.643486 undrain {"idset":"0"}
1589944129.337978 drain {"idset":"1","reason":"Meep foo"}
1589944147.935023 resource-init {"restart":true,"hwloc-discover":true,"online":"0-3","drain":"1","exclude":""}
```
Note from this example that hwloc discovery was skipped when the module was reloaded, and that the single node that was drained remained so after the reload.

**goals of this PR**
This PR establishes a general architecture for monitoring, and manipulating resource availability in a scheduler-neutral manner.

It enables parts of our near term milestone:
* ability drain/exclude nodes in the system instance
* ability to start system instance without all nodes online

**limitations**
The module proposed in this PR has shortcomings that we will want to address after that milestone like:
* replace resource object with new design
* resource object read from parent KVS in sub instance
* resource object read from config in system instance (checked with hwloc)
* handle resources not associated with execution target (or extend execution target concept beyond broker rank)
* finer grained drain/exclude than execution target (e.g. core)
* additional interfaces for tools

This PR does not come with a `flux resource` tool - @grondo has one prototyped on his branch that modifies `sched-simple` to use the `resource.acquire` PR.

A couple of lines need to be changed once flux-framework/flux-sched#655 (name change) is merged.

File: src/modules/resource/discover.c [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        flux_log_error (ctx->h, "hwloc.by_rank");
+    /* discover->f remains valid so by_rank can be accessed later */
+}
+
+static int lookup_hwloc (struct discover *discover)
+{
+    struct resource_ctx *ctx = discover->ctx;
+
+    assert (discover->f == NULL);
+    assert (discover->loaded);
```
Comment (dongahn) [inline src/modules/resource/discover.c:90]: Do we need this assert check? You probably have a good reason for it but since `assert()` will kill the entire flux instance, this caught my eye. 

By the way, in general when do we call terminating assert()? Like when it smells like a memory corruption?

**Comment by dongahn** [comment]
> This PR does not come with a flux resource tool - @grondo has one prototyped on his branch that modifies sched-simple to use the resource.acquire PR.

@grondo: could you post your prototyped sched-simple? This should give me a more concrete idea both flux-sched integration of this.

**Comment by dongahn** [comment]
> A couple of lines need to be changed once flux-framework/flux-sched#655 (name change) is merged.

FYI -- https://github.com/flux-framework/flux-sched/pull/655 was just merged.

**Comment by garlick** [comment]
For the record, this won't _break_ fluxion right now.   The KVS lookup using the WAITCREATE flag on hwloc XML files will continue to work.

Once we alter the broker startup so that rank 0 can start without the other ranks, then fluxion will still be able to acquire resources that way, but it will need to mark execution targets _down_ until they appear in an _up_ idset.  That will hopefully be coming soon.

After we sort out the new "resource object", then schedulers will need to have the ability to bootstrap from one of those.

File: src/modules/resource/discover.c (hunk 1 above)
Comment (garlick) [reply src/modules/resource/discover.c:90]: The intent was to document an invariant, but you're probably right, that's not a great idea in module/broker code.

**Comment by grondo** [comment]
> @grondo: could you post your prototyped sched-simple? This should give me a more concrete idea both flux-sched integration of this.

Yes, I will do this as I've just rebased on this branch. I'm still vacillating on the `sched.status` interface, but this is mainly just proof-of-concept at this point, and can be optional... but expect changes.

**Comment by garlick** [comment]
Just pushed some fixup commits to
* drop the assertion @dongahn pointed out
* rename `xresource` -> `resource`
* fix a leak found running sharness test under valgrind

**Comment by grondo** [comment]
Hm, looks like flux-sched check failed its valgrind check. I'm going to restart and see if it is reproducible.

**Comment by garlick** [comment]
(squashed the fixups per discussion on the coffee webex)

**Comment by grondo** [comment]
Hm, centos 8 builder failed here:
```
FAIL: t2800-jobs-cmd.t 42 - flux-jobs default color works (pty)
#	
#	        $runpty flux jobs --suppress-header --filter=pending | tail -n +2 > colorP_default.out &&
#	        check_no_color colorP_default.out &&
#	        $runpty flux jobs --suppress-header --filter=running | tail -n +2 > colorR_default.out &&
#	        check_no_color colorR_default.out &&
#	        $runpty flux jobs --suppress-header --filter=completed | tail -n +2 > colorCD_default.out &&
#	        count=`grep "\\u001b\[01\;32m" colorCD_default.out | wc -l` &&
#	        test $count -eq 4 &&
#	        $runpty flux jobs --suppress-header --filter=failed | tail -n +2 > colorF_default.out &&
#	        count=`grep "\\u001b\[01\;31m" colorF_default.out | wc -l` &&
#	        test $count -eq 1 &&
#	        $runpty flux jobs --suppress-header --filter=cancelled | tail -n +2 > colorCA_default.out &&
#	        count=`grep "\\u001b\[37m" colorCA_default.out | wc -l` &&
#	        test $count -eq 1
#	
```

@chu11, maybe this test still has some other small raciness?

**Comment by garlick** [comment]
The centos8 builder failed here (restarting):
```
not ok 42 - flux-jobs default color works (pty)
FAIL: t2800-jobs-cmd.t 42 - flux-jobs default color works (pty)
#	
#	        $runpty flux jobs --suppress-header --filter=pending | tail -n +2 > colorP_default.out &&
#	        check_no_color colorP_default.out &&
#	        $runpty flux jobs --suppress-header --filter=running | tail -n +2 > colorR_default.out &&
#	        check_no_color colorR_default.out &&
#	        $runpty flux jobs --suppress-header --filter=completed | tail -n +2 > colorCD_default.out &&
#	        count=`grep "\\u001b\[01\;32m" colorCD_default.out | wc -l` &&
#	        test $count -eq 4 &&
#	        $runpty flux jobs --suppress-header --filter=failed | tail -n +2 > colorF_default.out &&
#	        count=`grep "\\u001b\[01\;31m" colorF_default.out | wc -l` &&
#	        test $count -eq 1 &&
#	        $runpty flux jobs --suppress-header --filter=cancelled | tail -n +2 > colorCA_default.out &&
#	        count=`grep "\\u001b\[37m" colorCA_default.out | wc -l` &&
#	        test $count -eq 1
#	
```

**Comment by grondo** [comment]
> The centos8 builder failed here (restarting):

Sorry, I guess I failed to restart it!

**Comment by grondo** [comment]
ASan builder failed the same test:
```
020-05-20T22:21:20.1512030Z not ok 42 - flux-jobs default color works (pty)
2020-05-20T22:21:20.1512310Z FAIL: t2800-jobs-cmd.t 42 - flux-jobs default color works (pty)
2020-05-20T22:21:20.1512426Z #	
2020-05-20T22:21:20.1512892Z #	        $runpty flux jobs --suppress-header --filter=pending | tail -n +2 > colorP_default.out &&
2020-05-20T22:21:20.1513000Z #	        check_no_color colorP_default.out &&
2020-05-20T22:21:20.1513305Z #	        $runpty flux jobs --suppress-header --filter=running | tail -n +2 > colorR_default.out &&
2020-05-20T22:21:20.1513590Z #	        check_no_color colorR_default.out &&
2020-05-20T22:21:20.1514396Z #	        $runpty flux jobs --suppress-header --filter=completed | tail -n +2 > colorCD_default.out &&
2020-05-20T22:21:20.1514691Z #	        count=`grep "\u001b\[01\;32m" colorCD_default.out | wc -l` &&
2020-05-20T22:21:20.1514915Z #	        test $count -eq 4 &&
2020-05-20T22:21:20.1515218Z #	        $runpty flux jobs --suppress-header --filter=failed | tail -n +2 > colorF_default.out &&
2020-05-20T22:21:20.1515868Z #	        count=`grep "\u001b\[01\;31m" colorF_default.out | wc -l` &&
2020-05-20T22:21:20.1516141Z #	        test $count -eq 1 &&
2020-05-20T22:21:20.1516439Z #	        $runpty flux jobs --suppress-header --filter=cancelled | tail -n +2 > colorCA_default.out &&
2020-05-20T22:21:20.1516733Z #	        count=`grep "\u001b\[37m" colorCA_default.out | wc -l` &&
2020-05-20T22:21:20.1516979Z #	        test $count -eq 1
2020-05-20T22:21:20.1517086Z #	
```

Will restart

**Comment by garlick** [comment]
Rebased on current master to pick up @chu11's test fix.

**Comment by grondo** [comment]
I hit one issue with this PR. In retrospect it should be obvious, but I bring it up in case there is some mitigation.

If the `resource` module is reloaded, this obviously terminates the `resource.acquire` streaming RPC, but the scheduler is not notified of this (in any way I'm aware of), so it sits waiting on resource updates that will never come. Meanwhile, the newly reloaded resource module thinks there is no `resource.acquire` client to which to send idset updates, so `drain` and `undrain` service still works, but updates are not reflected in scheduler resources.

Not sure if there's a good way to handle this, the only ideas I had were:

 * `resource` module refuses to exit if an `acquire` streaming RPC is "in progress" (i.e. force scheduler to be unloaded before `resource`)

 * `resource` module sends a `restarted` event that `acquire` users subscribe to, which would tell scheduler to restart the acquire protocol.

**Comment by garlick** [comment]
Oh, I thought I had handled that by sending ECANCELED response to the scheduler if an RPC was active during shutdown, but you are right I missed that!  Will fix.

If the scheduler wants to recover from that, it could, but it seems like it is fair game to treat that as a fatal error.  There's an implied loading order here, even if it's not explicitly enforced.  For example, if the scheduler is loaded before the resource module it will get an ENOSYS response to acquire.

**Comment by grondo** [comment]
> If the scheduler wants to recover from that, it could, but it seems like it is fair game to treat that as a fatal error. There's an implied loading order here, even if it's not explicitly enforced. For example, if the scheduler is loaded before the resource module it will get an ENOSYS response to acquire.

Yeah, it isn't possible to recover unless the resource module was able to announce its availability on reload (scheduler won't know when it is able to "re-acquire" resources). So I think you're right, the scheduler should just exit with error if it gets an error from `acquire` RPC. At least the system won't get into an inconsistent state.

**Comment by garlick** [comment]
Oh you're right, I misspoke on recovery.

**Comment by grondo** [comment]
BTW, there is one other good case for making a module un-unloadable. I accidentally unloaded connector-local once and _that_ didn't go well... :wink:

**Comment by garlick** [comment]
Start your own flux instance - with new and improved foot gun!

**Comment by garlick** [comment]
Just pushed a fix (squashed).

**Comment by grondo** [comment]
Great, that works:
```console
$ flux module remove resource
2020-05-22T22:01:35.174629Z sched-simple.err[0]: exiting due to resource update failure: Operation canceled
$
```

**Comment by grondo** [comment]
> Great, that works

Though this does imply that scheduler should always be unloaded *before* resource (currently in this PR it is the other way around)

```
 grondo@asp:~/git/flux-core.git$ grep -C1 resource etc/rc3 

flux module remove -f resource
flux module remove -f sched-simple
```

**Comment by garlick** [comment]
Ah, good catch.  I'll fix that too (and also fix up the merge conflict).

**Comment by garlick** [comment]
Note I did return a textual error: "the resource module was unloaded", so might be good to print that if available.

**Comment by grondo** [comment]
Github seems to be having intermittent issues today. I restarted the workflows since they failed early while downloading actions.

**Comment by garlick** [comment]
Just restarted the chain lint bionic builder - it also seemed to fail before it really got going.

**Comment by grondo** [comment]
I'm seeing the following failure on a TOSS system with 36 cpus:

```
expecting success: 
	flux kvs get resource.hwloc.by_rank >/dev/null &&
	seq 0 $(($SIZE-1)) >hwloc_xml.exp &&
	flux kvs ls -1 resource.hwloc.xml >hwloc_xml.out &&
	test_cmp hwloc_xml.exp hwloc_xml.out

--- hwloc_xml.exp	2020-05-23 01:39:29.924426000 +0000
+++ hwloc_xml.out	2020-05-23 01:39:29.980421000 +0000
@@ -1,5 +1,12 @@
 0
 1
+10
+11
+12
+13
+14
+15
+16
 2
 3
 4
@@ -8,10 +15,3 @@
 7
 8
 9
-10
-11
-12
-13
-14
-15
-16
not ok 8 - resource.hwloc is populated after hwloc-discover-finish
#	
#		flux kvs get resource.hwloc.by_rank >/dev/null &&
#		seq 0 $(($SIZE-1)) >hwloc_xml.exp &&
#		flux kvs ls -1 resource.hwloc.xml >hwloc_xml.out &&
#		test_cmp hwloc_xml.exp hwloc_xml.out
#	
```

**Comment by garlick** [comment]
Doh!  I need to sort that list!

**Comment by garlick** [comment]
There I think that's got it.

**Comment by grondo** [comment]
@garlick, I'm probably doing something wrong, but I thought I'd post here before I start pulling my hair out.

In my branch with sched-simple converted to use resource.acquire, I'm getting occasional hangs loading `sched-simple` because there is no initial response to the resource.acquire RPC.

Adding a little debug, it looks like this occurs when the request comes before `lookup_hwloc` is complete, so the "defer response" branch is taken.

When I check the resource.eventlog in a successfully started instance, I don't see the "discover" event that seems to trigger the first response in `acquire.c:reslog_cb()` (If I'm reading things right), only 
```
1590202558.240104 hwloc-discover-start {"pid":30576}
1590202558.438961 hwloc-discover-finish {"loaded":true}
```

Is there an emitted event missing?

**Comment by garlick** [comment]
Uh oh, maybe I didn't rename the triggering event name to
hwloc-resource-finish when I changed the event names at the source?  Will
have a look tomorrow morning.

**Comment by dongahn** [comment]
> Yeah, it isn't possible to recover unless the resource module was able to announce its availability on reload (scheduler won't know when it is able to "re-acquire" resources). So I think you're right, the scheduler should just exit with error if it gets an error from acquire RPC. At least the system won't get into an inconsistent state.

Sorry I'm coming at this late (I took one day vac today).

One thing that might be helpful is to pencil in the overall failure and failover semantics across different scheduler-related modules somewhere: job-manager, resource, and schedulers.

For example, I built the state reconstruction semantics of `sched-fluxion-qmanager` assuming job-manager resiliency. This made it easy to reason about the resiliency of `sched-fluxion-qmanager`.

I will also need to add a similar semantics for `sched-fluxion-resource` w/ respect to `resource` (as well as `sched-fluxion-qmanager`). I don't think there is any issue here though.

Now one thing that I don't think I have heard much is the proper semantics of the schedulers when `job-manager` and/or `resource` fails (short term: when they are shut down).

I'm just saying this because with the two module approach of fluxion, I struggled with it a bit already (Please see the table in https://github.com/flux-framework/flux-sched/pull/625#issue-390749706). And we have one more participating module, so this space can become quickly more complex.

**Comment by dongahn** [comment]
BTW, do we expect any problem when `resource` is shut down and come back up in terms of recovering the resource state? Probably not an issue with the short term requirement where it can be assumed that the module will be shutdown when all compute nodes were brought up on line...

**Comment by garlick** [comment]
Just pushed a fix for the problem @grondo found (needed to update the event name - sorry bout that!)

@dongahn, the resource module keeps its state in `resource.eventlog` which it replays when it reloads.  This allows it to 1) keep track of nodes that were drained at runtime across module reload, and 2) track whether hwloc data has been fetched, to avoid repeating a costly operation.  The main use case this supports is system instance restart, where KVS state, and thus the resource eventlog, is preserved across a cluster reboot or systemd restart.

Another supported use case is reloading the scheduler.  The resource module receives a `resource.disconnect` notification when the scheduler unloads so it can terminate the `resource.acquire`  stream and free up the slot.  This includes if the scheduler were to "panic" and exit its `mod_main()`, or if the instance owner wants to unload the scheduler and reload a different one.  IMHO the scheduler is the most challenging design in the system, and also the most likely to be replaced/reconfigured by the user, so it is loaded _last_ so that it has maximum recoverability.

We don't support unloading the resource module while the scheduler is loaded, though we try to fail gracefully.  When this happens, `resource.acquire` will receive an ECANCELED error and, as discussed above, there is not much the scheduler can do to recover (log and exit `mod_main()` is probably the best strategy).  If we need to recover from that, we could develop protocols for it, but right now the only way that could happen is if the resource manager internally fails, or the instance owner unloads the resource module out of order, somewhat unlikely IMHO.

(It's another story in test, where we unload and reload modules frequently - we just must respect the implicit loading order)

Does that answer your question?

**Comment by garlick** [comment]
(restarted a couple of builders that failed to download stuff)

**Comment by grondo** [comment]
> Just pushed a fix for the problem @grondo found (needed to update the event name - sorry bout that!)

Thanks! That works.

But there isn't still a small race because `hwloc-discover-finish` is posted after `flux hwloc reload` finishes, but before `flux_kvs_lookup` of `resource.hwloc.by_rank` is complete? (Sorry if I've misunderstood the code!)

**Comment by garlick** [comment]
> But there isn't still a small race

If `discover.get()` is called before the lookup completes, it just blocks until it does. Blocking it is undesirable but I figured it's no big deal in this case since 1) it's a one time thing, 2) it should be quick (milliseconds tops)

There should be no possibility for `discover_get()` to be called before the KVS lookup has _started_.

**Comment by grondo** [comment]
> If discover.get() is called before the lookup completes, it just blocks until it does. Blocking it is undesirable but I figured it's no big deal in this case since 1) it's a one time thing, 2) it should be quick (milliseconds tops)

Thanks, that makes sense! (also, not much else will be going on in the resource module at this time)

**Comment by garlick** [comment]
Oh: I forgot, the other mitigating factor is that the callback isn't made until the eventlog commit _completes_, but the KVS lookup starts when the hwloc subprocess completes and the eventlog commit _starts_.  So the two KVS ops are taking place in parallel.

File: src/modules/resource/acquire.c [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+ *
+ * First response:
+ *   {resources:resource_object up:idset}
+ * Subsequent responses:
+ *   {up?:idset down?:idset}
+ *
+ * Where:
+ * - resource_object maps execution target id's to resources
+ *   (see RESOURCE OBJECT) below
+ * - idset is a set of execution target id's, encoded as a string.
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/modules/resource/acquire.c:25]: I keep doing that!  Grr.  Thanks, will fix.

File: src/modules/resource/drain.c [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                idset_destroy (idset);
+            }
+        }
+    }
+    return 0;
+}
+
+static const struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST,  MODULE_NAME ".drain", drain_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST,  MODULE_NAME ".undrain", undrain_cb, 0 },
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/modules/resource/drain.c:264]: That would have been fine.  If many messages were going to be parsed by the module, there might be a small advantage to avoiding the glob, since `dispatch_message()` (internal to msg_handler.c) uses a hash to look up request handlers by topic string, then if that fails, it does a linear search through a list where globs are matched with `flux_msg_cmp()`.

File: t/t2310-resource-module.t [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	EOT
+	flux config reload &&
+	$WAITFILE -t 10 -v -p \"down\" acquire5.out &&
+	cat >resource.toml <<-EOT &&
+	[$RESNAME]
+	exclude = "0"
+	EOT
+	flux config reload &&
+	$WAITFILE -t 10 -v -p \"up\" -c 2 acquire5.out &&
+	kill -15 $pid && wait $pid || :
```
[the remark this answers is not in the document]
Comment (garlick) [reply t/t2310-resource-module.t:252]: I would not want to invite a scathing reivew from @chu11!

**Comment by garlick** [comment]
Forced a push with the suggested changes.  Thanks!

**Comment by grondo** [comment]
Feel free to set MWP then!

**Comment by garlick** [comment]
Looks like the ASAN builder was canceled after 40m (but I can't find any smoking gun as to why).  Restarting.

**Comment by dongahn** [comment]
@garlick: thank you for the detailed explanation. 

>  IMHO the scheduler is the most challenging design in the system, and also the most likely to be replaced/reconfigured by the user, so it is loaded last so that it has maximum recoverability.

I agree. 

> We don't support unloading the resource module while the scheduler is loaded, though we try to fail gracefully. When this happens, resource.acquire will receive an ECANCELED error and, as discussed above, there is not much the scheduler can do to recover (log and exit mod_main() is probably the best strategy).

Thanks. Yes this makes sense. The "live" recovery will require a lot more logic.  I agree it is cleaner to do a reload-based recovery (-- live recovery will be in a form of waiting for `resource` to be reloaded -- we don't save much). 

In general, as far as `job-manager` and/or `resource` can be reloaded and reconstruct its state, the schedulers should be able to recover their states through 'reload'. 

> If we need to recover from that, we could develop protocols for it, but right now the only way that could happen is if the resource module internally fails, or the instance owner unloads the resource module out of order, somewhat unlikely IMHO.

> (It's another story in test, where we unload and reload modules frequently - we just must respect the implicit loading order)

> Does that answer your question?

All of these make sense and are consistent with my understanding.

I just thought that something written such that scheduler failure/failover semantics and internals can be checked against that contract. And more rigorous testing can be designed that way too.

Do you think it is worthwhile to do a RFC on the semantics (including proper loading order and such)? 

The verbal exchange in this PR can serve as our semantics in the meantime though. Maybe we should circle back to this after our next major milestone.

**Comment by garlick** [comment]
I certainly think that documenting the job-manager / scheduler interface, as well as this one (maybe after we stablize it a bit?) makes a lot of sense.  The "recovery semantics" could be part of such RFC(s).

File: t/t2310-resource-module.t (hunk 4 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply t/t2310-resource-module.t]: `:` is at least faster!
